### DISCUSSION TITLE
How to think about granularity

### DISCUSSION THREAD

**Discussion by jasnonaz** [dbt-labs/devhub-writing#73] [Analytics craft]
In @lbenezra-FA's fantastic post on [joining snapshots](https://github.com/dbt-labs/dbt-technical-blog-writing/issues/56), there is line that reads

> Step 4: Fan out your data tables to the grain necessary for the resulting tables. 
What happens in this step? Step 4 walks you through how to do your first join, in which you need to fan out the data spine to the finest grain possible. This step is crucial to joining the snapshots later steps. 

This got me thinking about the concept of granularity in data - how to identify the granularity of a dataset, select the appropriate granularity and what this means for how models should be built. This is a hugely important topic that gets _referenced_ a lot, but I'm not sure there exists a good resource to learn about granularity. Curious if others have thoughts on how we might create resources for learning about granularity.

**Comment by amychen1776** [comment]
I love this. When I first got started, understanding what granularity was so confusing. I have countless conversations with Jerco going over why it was key to know the primary key and how to find out what fanout was because it was this "granularity" that I didn't want. What's funny is most of the texts/online resources that I have that mentions granularity are always just a paragraph or so long. I would love a resource on how to identify the grain of a table and then how to fix ways to see what broke (i.e. not just count(primary_key))

**Comment by patkearns10** [comment]
[Kimball](https://www.kimballgroup.com/2007/07/keep-to-the-grain-in-dimensional-modeling/) always does a good job explaining core concepts, like grain.

some thoughts for dbt at the staging layer:
- objects from an operational database are (typically) very straightforward and easy to grok -- an `orders` table likely has an id for each order, meaning the grain is "one row for each order placed". Generally primary keys are defined in the raw database as well, so oftentimes you can look up what the grain of that table is already in the db.
- objects from NOSQL databases, like MongoDB are inherently more difficult, because they give you nested json tables which could be unpacked into multiple grains (many to many), and contain possible duplicates. There is often no constraints on the tables, which means you cant be told what the grain is, but there is usually good documentation (somewhere) that shows what the structure of the json should look like (and what could possibly end up in the data), but typically you will have to do some exploratory data analysis to really understand it.  (this dovetails nicely to @christineberger 's json post)

As you start building models (derived, granular levels) you start combining data from your staging with a purpose.  Here you should have an idea of the grain you are trying to create, as opposed to just trying to surface what is the grain of the data that was given to you. 
- You might be keeping the grain from above from orders and building a `fct_orders` table, where you are simply joining more tables and adding additional dims or measures to it, in which case, ensure everything you do keeps the grain of the table at the order_id grain and you are not accidentally joining in tables that have multiple rows per each order id (like order items table), which would cause the `fct_orders` table to fan out.
- or you might be creating a new specific grain for a new analysis, like a `kpi_metrics` table that snapshots your kpis daily. So here, you would expect the grain of the table to be `date+metric`, meaning you will have 1 row for each metric that existed that day the snapshot occurred.  As you add or remove metrics at your company, those metric values will change over time, but you should have at least one metric for every day (unless of course you remove them all on purpose, or if your pipeline breaks and you dont fix it within a day or backfill the missing data).

**Comment by patkearns10** [comment]
Also, worth mentioning that my most common ad hoc query of all time is: 
```
select 
  supposedly_unique_id,   -- # or concat(id_1, id_2, ...)
  count(*)
from
  table_to_add_primary_key_tests
group by 1
having count(*) > 1
```
I have always wondered if there was an optimized way to crawl through your models and determine primary keys of tables, and use that to both create primary key tests and also confirm what you thought was the grain is in fact the grain of the table.

**Comment by CodingSanTu** [comment]
你的邮件我已收到--by三土
